compiler/src/iree/compiler/Dialect/Flow/Transforms/FusionPreprocessing.cpp

```diff
@@ -131,14 +142,78 @@ struct FoldSuccessiveTensorInsertSliceOps
   }
 };
 
+//===----------------------------------------------------------------------===//
+// GatherFusionPattern
+//===----------------------------------------------------------------------===//
+
+// Specific case. The linalg generic implementation of "gather"
+// cannot be fused because it there is no producer-consumer
+// relationship between the two generics. This is because the indexing
+// is not affine (index values come from a tensor).
+struct GatherFusionPattern : public OpRewritePattern<tensor::ExtractOp> {
+  using OpRewritePattern<tensor::ExtractOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(tensor::ExtractOp extractOp,
+                                PatternRewriter &rewriter) const override {
+    // Check if extractOp is inside a generic op
+    auto consumerOp =
+        dyn_cast_or_null<linalg::GenericOp>(extractOp->getParentOp());
+    if (!consumerOp) {
+      return rewriter.notifyMatchFailure(
+          extractOp, "expected extract op to be inside a generic op");
+    }
+
+    auto producerOp = dyn_cast_or_null<linalg::GenericOp>(
+        extractOp.getOperand(0).getDefiningOp());
```

Comment:
nit: Move the casting to `getDefinigOp<>` and use the accessor from `tensor::ExtractOp` (i.e., .`getTensor()`. So this could be `extractOp.getTensor().getDefiningOp<linalg::GenericOp>())`